### PR TITLE
Fetch data for all PyPI packages

### PR DESCRIPTION
**Pull request by hugovk**
Fixes https://github.com/hugovk/top-pypi-packages/issues/36.

It costs the same to get download data for _all_ ~589,368 packages on PyPI as for 8,000 or for 5,000 or for 1 package!

https://dev.to/hugovk/a-surprising-thing-about-pypis-bigquery-data-2g9o

So let's fetch data for them all!

The JSON for all packages is 31 MB and CSV is 12 MB, so let's zip those to 5.2 MB and 4.4 MB.

Let's still provide uncompressed JSON and CSV for 8,000 packages.

### CODE REVIEW

File: generate.sh [hunk 1]
```diff
@@ -15,8 +15,20 @@ python3 -m pip install -U pypinfo
 python3 -m pip --version
 /home/botuser/.local/bin/pypinfo --version
 
+# Check if zip is installed
+if ! command -v zip &> /dev/null
+then
+    echo "zip not be found, consider: apt install zip"
```
Comment (hugovk) [inline generate.sh:21]:
```suggestion
    echo "zip not found, consider: apt install zip"
```

**Comment by lesteve** [comment]
I bumped into your blog post, super interesting thanks (I originally came to your blog for setup-python free-threaded support)! Thanks as well for this repo, I found it super useful in the past, in particular to get some data without having to worry about running Google BigQuery queries, free quota and all :pray:!

I completely get the operational cost thing and that it is more convenient to take all downloads (rather than only `installer='pip'`) and that fetching the data for all the packages costs the same so you may as well provide the data for all PyPI packages. 

Having said this, I am a bit worried that for packages that are downloaded less, people will start trusting these numbers blindly without being aware of the caveats (most downloads are from mirrors and not `pip` or `uv` installs). Even for packages that are downloaded more, the difference can be 10% which is not small.

For example recently I had a closer look at [skore](https://github.com/probabl-ai/skore) downloads and out of the 12K downloads in the last month, only 1.5k (so 12%) were from `pip` or `uv`. 

Full disclosure: I have an upcoming PyCon Italia talk about some of the PyPI download stats caveats I have discovered along my journey, here is the [abstract](https://2025.pycon.it/it/event/pypi-in-the-face-running-jokes-that-pypi-download-stats-can-play-on-you).

**Comment by hugovk** [comment]
> I bumped into your blog post, super interesting thanks (I originally came to your blog for setup-python free-threaded support)! Thanks as well for this repo, I found it super useful in the past, in particular to get some data without having to worry about running Google BigQuery queries, free quota and all 🙏!

Thanks!


> I completely get the operational cost thing and that it is more convenient to take all downloads (rather than only `installer='pip'`) and that fetching the data for all the packages costs the same so you may as well provide the data for all PyPI packages.

By the way, I tried fetching all but for some reason the process was killed on the server, so I moved it back down to 15k (still up from 8k): https://github.com/hugovk/top-pypi-packages/pull/44. I might try increasing it again in the future...


> Having said this, I am a bit worried that for packages that are downloaded less, people will start trusting these numbers blindly without being aware of the caveats (most downloads are from mirrors and not `pip` or `uv` installs). Even for packages that are downloaded more, the difference can be 10% which is not small.
> 
> For example recently I had a closer look at [skore](https://github.com/probabl-ai/skore?rgh-link-date=2025-04-02T07%3A51%3A36Z) downloads and out of the 12K downloads in the last month, only 1.5k (so 12%) were from `pip` or `uv`.

Yeah, there's always trade-offs, and that's a good point about less-downloaded packages.

Let's compare skore with the package currently at rank 100 (referencing):

```console
❯ pypinfo --markdown --all --percent --start-date 2025-03 --end-date 2025-03 skore installer
Served from cache: False
Data processed: 4.18 GiB
Data billed: 4.18 GiB
Estimated cost: $0.03
```
| installer_name | percent | download_count |
| -------------- | ------: | -------------: |
| bandersnatch   |  34.32% |            775 |
| requests       |  18.11% |            409 |
| uv             |  15.32% |            346 |
| pip            |  13.55% |            306 |
| Browser        |  10.27% |            232 |
| None           |   7.66% |            173 |
| conda          |   0.62% |             14 |
| poetry         |   0.13% |              3 |
| Total          |         |          2,258 |


```console
❯ pypinfo --markdown --all --percent --start-date 2025-03 --end-date 2025-03 referencing installer
Served from cache: False
Data processed: 3.50 GiB
Data billed: 3.50 GiB
Estimated cost: $0.02
```
| installer_name | percent | download_count |
| -------------- | ------: | -------------: |
| pip            |  77.01% |     95,788,632 |
| uv             |  16.35% |     20,336,635 |
| poetry         |   4.80% |      5,974,879 |
| requests       |   1.44% |      1,793,027 |
| None           |   0.29% |        355,183 |
| Nexus          |   0.05% |         62,962 |
| Browser        |   0.02% |         28,823 |
| Bazel          |   0.02% |         25,304 |
| pdm            |   0.01% |         11,603 |
| bandersnatch   |   0.00% |          4,624 |
| Total          |         |    124,381,672 |

That's 33% uv+pip for skore and for 93% uv+pip referencing.

I expect we'll get a similar high uv+pip share near the top, and similar high mirror share near the bottom? And that the share will change fairly smoothly as we go down the list?

If so, in any given month, I don't think it should affect relative positioning too much? Those ranking near the top will have similar numbers and a similar share, and those near skore will also have similar numbers and a similar share. There'll of course be some outliers.

Comparing absolute numbers one month to another isn't necessarily a good idea, because of the changes I unfortunately need to make to stay within quota. I've tried to list them in the changelog for visibility: https://hugovk.github.io/top-pypi-packages/#changelog.

However, I usually see people using this data when they want to do some study on the most popular packages (see https://hugovk.github.io/top-pypi-packages/#users), so these relative numbers are hopefully useful as a recent accessible snapshot.

> Full disclosure: I have an upcoming PyCon Italia talk about some of the PyPI download stats caveats I have discovered along my journey, here is the [abstract](https://2025.pycon.it/it/event/pypi-in-the-face-running-jokes-that-pypi-download-stats-can-play-on-you).

Looks interesting! I'll be at PyCon Italia too :)

**Comment by lesteve** [comment]
Thanks for your thoughts on this! I agree it's all about trade-offs and I guess since you are doing the top 15k packages I am guessing that it's not too bad.

> Looks interesting! I'll be at PyCon Italia too :)

Great to hear! Basically the talk proposal was prompted by being mildly frustrated by people looking at PyPI download stats and telling stories with it without being aware (or caring much if I am more pessimistic) about the caveats.

Maybe I missed them but I was not able to find resources explaining the complexity of the reality behind those seemingly simple numbers, so I thought oh well let's try to transform this mild frustration into a talk then :sweat_smile:!

**Pull request merged**